docs/proposals/20250724-independent-worker-topology-versions.md

```diff
@@ -0,0 +1,499 @@
+of ClusterClass while introducing operational flexibility needed for production environments at scale.
+
+The broader motivation aligns with the goals of Cluster API: declarative infrastructure, safe upgrades, and supporting real-world
+platform needs without requiring users to abandon upstream tools or adopt brittle workarounds.
+
+Optionally, future experience reports can be gathered from teams running large-scale Kubernetes fleets with mixed workloads
+(e.g., inference vs training, customer A/B rollout clusters) to further validate this direction.
+
+
+### Goals
+
+- Support upgrades of any single worker node group independent of the control plane and other worker node groups.
+- Operations on worker node groups *not* upgrading should continue to operate normally while an independent worker node group is upgrading.
+- Support version skew, as allowed by Kubernetes' [Version Skew Policy](https://kubernetes.io/releases/version-skew-policy/), between worker node groups and control planes. 
+- Maintain supported skew checks and guardrails for Kubernetes versions
+- Maintain the ease of use of ClusterClass while providing this added flexibility
+- Support switching back to cluster-level Kubernetes versions if all independent worker node groups and the control plane are at the same version.
+- Gracefully handle [version-aware patches](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/write-clusterclass#version-aware-patches) (i.e.  ClusterClass variables can reference `builtin.cluster.topology.version`, `builtin.controlPlane.version`, `builtin.machineDeployment.version`, and `builtin.machinePool.version`)
+- Update documentation in the [cluster-api book](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/operate-cluster#upgrade-a-cluster) to describe the alternative approach to upgrading cluster versions.
+
+### Non-Goals
+
+#### Switching between cluster-level and independent Kubernetes versions when versions don't match
+To minimize potential edges cases, if a cluster vendor wishes to use worker node group versioning, they cannot switch back without ensuring that the independent version matches the version that is the cluster.spec.topology version. Otherwise, they will be denied and asked to take manual steps to get the control versions the same. 
+
+#### Allowing version skews that are unsupported by kubernetes
+Only version skew allowable by Kubernetes [Version Skew Policy](https://kubernetes.io/releases/version-skew-policy/) can be implemented. 
+
+#### Changing behavior of cluster-level worker node group configuration rollouts
+Rollouts for cluster-level behavior will be preserved. 
+
+#### Add builtin version-aware patches for independent worker node groups
+For this iteration, we will not be implementing version-aware patches for independent worker node groups. 
+
+### Future Work
+
+- Support flipping between cluster-level version control and independent worker version control modes. 
```

Comment:
```suggestion
- Allow removal of worker group version overrides when all worker nodes can be upgraded to match the control plane’s version. For example, if the control plane is running 1.31.x and all independent worker node groups are at 1.30.x, a cluster vendor can eliminate the worker group version specifications. This will trigger all worker groups to upgrade to 1.31.x and effectively revert to cluster-level version management.
```